# Wrangling + Tf-Idf model

In [1]:
import pandas as pd

import warnings
warnings.filterwarnings('ignore')

In [2]:
#data = pd.read_csv('data.csv', error_bad_lines=False, encoding="utf-8") # CSV old way
data = pd.read_json("data_5scheduler.json") # json new way

Look at the data:

In [3]:
print(data.describe(include="object"))

                title   identifier description  source instructors offered  \
count            4443         4443        4443    4443        4443    4443   
unique           3761         4234        3989       5        1231     139   
top     Senior Thesis  PHYS-178-KS              Pomona          []           
freq               56            2          95    1446        2023    1559   

       prerequisites corequisites  
count           4443         4443  
unique           716           28  
top                                
freq            3359         4411  


### Duplicates:
It looks like there is only 3989 unique course descripition so let's remove duplicates based on 'description' column.
There are also rows with empty descriptions, which are not helpful

In [4]:
print(len(data))
data = data.drop_duplicates(subset='description')
data = data.drop_duplicates(subset='identifier')
data = data[data["description"] != ""]
data.reset_index(inplace=True,drop=True)
print(len(data))

4443
3853


In [5]:
data.head(7)

title   identifier  \
0                  Introduction to American Cultures  AMST-103-HM   
1                         Print and American Culture  AMST-115-HM   
2                               Hyphenated Americans  AMST-120-HM   
3    Life: Knowledge, Belief, and Cultural Practices  ANTH-110-HM   
4  Introduction to the Anthropology of Science an...  ANTH-111-HM   
5                                   War and Conflict  ANTH-115-HM   
6                                      Rationalities  ANTH-134-HM   

                                         description      source  credits  \
0  An interdisciplinary introduction to principal...  HarveyMudd      300   
1  Covers numerous developments in American print...  HarveyMudd      300   
2  A focus on the experience of immigrants in the...  HarveyMudd      300   
3  An exploration of cultural attitudes toward li...  HarveyMudd      300   
4  An introduction to science and technology as c...  HarveyMudd      300   
5  “The wings of the butterfly—that cause the hur...  HarveyMudd      300   
6  What does it mean to be rational? Does it mean...  HarveyMudd      300   

          instructors                  offered  \
0             [Staff]                            
1        [Anup Gampa]                            
2          [Balseiro]                            
3           [de Laet]                            
4  [Marianne De Laet]                            
5           [de Laet]                            
6           [de Laet]  Offered alternate years   

                                       prerequisites corequisites  \
0                                                                   
1                                                                   
2                                                                   
3                                                                   
4                                                                   
5                                                                   
6  Any introductory course in anthropology or any...                

   currently_offered  fee  
0              False    0  
1               True    0  
2              False    0  
3              False    0  
4               True    0  
5              False    0  
6              False    0

### Tf-Idf with scikit-learn
[Description](https://monkeylearn.com/blog/what-is-tf-idf/)

[Usage](https://kavita-ganesan.com/tfidftransformer-tfidfvectorizer-usage-differences/#.Y1M42ezMJhF)

Here is an example of how Tf-Idf would work if our documents were the following 4 sentences:

In [6]:
import sklearn
from sklearn.feature_extraction.text import TfidfVectorizer
import numpy as np

corpus = [
     'this is the first document',
     'this document is the second document',
     'and this is the third one',
     'is this the first document',
]
vectorizer = TfidfVectorizer(use_idf=True)
vectors = vectorizer.fit_transform(corpus)
firstv = vectors[0]
df = pd.DataFrame(firstv.T.todense(), index=vectorizer.get_feature_names(), columns=["tfidf"])
df = df.sort_values(by=["tfidf"], ascending = False)
print("TfIdf values for the first sentence")
print(df)


TfIdf values for the first sentence
             tfidf
first     0.580286
document  0.469791
is        0.384085
the       0.384085
this      0.384085
and       0.000000
one       0.000000
second    0.000000
third     0.000000


In the example above we can see the importance of each word ranked for the first sentence `'this is the first document'`. So, for example the word `first` is important since it doesn't appear in any other document. The word `the` is not as important since it appears in all other documents. And the word `third` is not important at all since it doesn't even appear in the first document.

### Rank classes based on a given word
Function `tfidf_word(word, data)` takes in the word we are interested in and the data we are looking at. The function returns an updated dataframe with a new column `"score"` that gives each class a score of importance based on the input word. 

In [7]:
def tfidf_word(word, data_1):
    data_2 = data_1.copy() # since we don't want to be making changes to our original dataframe
    corpus = list(data_2.description)
    vectorizer = TfidfVectorizer(use_idf=True)
    vectors = vectorizer.fit_transform(corpus)

    score_for_word = []
    words = vectorizer.get_feature_names()
    try:
        index = words.index(word)
    except:
        print("'" + word + "'" + " is not mentioned in any course descriptions")
        return

    for i in range(0, len(corpus)):
        value = vectors[i].T.todense()[index]
        score_for_word.append(value)

    score_for_word = [float(i) for i in score_for_word] # type cast each score to a float

    data_2["score"] = score_for_word
    return data_2

For example, let's say we are interested in ranking all of the classes based on the word `computer`:

In [8]:
tfidf_word('computer', data).sort_values(by=["score"], ascending = False).head(10)

title     identifier  \
129                       Computer Science Seminar    CSCI-181-HM   
506             Special Topics in Computer Science    CSCI-181-CM   
1415                   Computer Science Colloquium    CSCI-188-PO   
413     Introduction to Computational Neuroscience   BIOL-133L-KS   
1416               Computer Science Senior Seminar    CSCI-190-PO   
1027         Computational Physics and Engineering    PHYS-100-KS   
499               Fundamentals of Computer Science    CSCI-052-CM   
133                    Computer Science Colloquium    CSCI-195-HM   
103   Introduction to Biology and Computer Science  CSCI-005GR-HM   
496                 Data Structures and Algorithms    CSCI-046-CM   

                                            description            source  \
129   Advanced topics of current interest in compute...        HarveyMudd   
506   Selected topics in computer science. May be re...  ClaremontMckenna   
1415  Colloquium presentations and discussions of to...            Pomona   
413   This course provides computational skills for ...  ClaremontMckenna   
1416  Reading, discussion and presentation of resear...            Pomona   
1027  This course is a comprehensive introduction to...  ClaremontMckenna   
499   A solid foundation in functional programming, ...  ClaremontMckenna   
133   Oral presentations and discussions of selected...        HarveyMudd   
103   This course introduces fundamental concepts fr...        HarveyMudd   
496   An introduction to abstract data types and fun...  ClaremontMckenna   

      credits           instructors          offered  \
129         0               [Staff]  Fall and Spring   
506       100                    []     Occasionally   
1415        0     [Joseph C Osborn]   Each semester.   
413       100                    []       Every fall   
1416       25     [Joseph C Osborn]   Each semester.   
1027      100          [Scot Gould]     Every spring   
499       100                    []     Occasionally   
133        50  [Melissa E. O'Neill]  Fall and Spring   
103       300  [Wu, Bush (Biology)]             Fall   
496       100          [Mark Huber]       Every year   

                                          prerequisites corequisites  \
129                            Permission of instructor                
506                                                                    
1415  CSCI 051A PO , or CSCI 051G PO , or CSCI 051J ...                
413                                                                    
1416  Senior standing and two CSCI core courses (inc...                
1027                                                                   
499                                                                    
133                            Juniors and seniors only                
103                                                                    
496                                                                    

      currently_offered  fee     score  
129               False    0  0.441565  
506               False    0  0.430887  
1415               True    0  0.421235  
413               False    0  0.338217  
1416               True    0  0.334543  
1027               True    0  0.325196  
499               False    0  0.315243  
133                True    0  0.277064  
103               False    0  0.256092  
496                True    0  0.247691

These are the first 10 instances of the classes that are most related to the word `computer` ranked in descending order (more related classes are on top). So, we could recomend a student who is interested in `computer`
 to take these classes.

Bellow are the outputs for fords `data, culture, activism, fiction, environment`

In [9]:
tfidf_word('data', data).sort_values(by=["score"], ascending = False).head(10)

title   identifier  \
507                   Advanced Projects in Data Science    DS-180-CM   
1303  Data Analysis and Programming for the Life Sci...  BIOL-174-PO   
494                         Foundations of Data Science  CSCI-036-CM   
258                            Nonlinear Data Analytics  MATH-178-HM   
2848                       CS1: Intro to Python and Viz    MS-059-SC   
2596                                       Econometrics  ECON-125-SC   
561                           Accounting Data Analytics  ECON-160-CM   
107             Data Structures and Program Development  CSCI-070-HM   
504                         Introduction to Data Mining  CSCI-145-CM   
941                         Introduction to Data Mining  MATH-166-CM   

                                            description            source  \
507   This course allows teams of students to wrestl...  ClaremontMckenna   
1303  This course explores the analysis of big data ...            Pomona   
494   Data science is the interdisciplinary study of...  ClaremontMckenna   
258   Analysis of nonlinear large dynamic data inclu...        HarveyMudd   
2848  This is an introduction to computer programmin...           Scripps   
2596  Statistical techniques for testing economic mo...           Scripps   
561   This course will introduce students to the use...  ClaremontMckenna   
107   Abstract data types including priority queues ...        HarveyMudd   
504   Data mining is the process of discovering patt...  ClaremontMckenna   
941   Data mining is the process of discovering patt...  ClaremontMckenna   

      credits                          instructors                    offered  \
507       100                          [Jeho Park]                 Every year   
1303      100                   [Andre Cavalcanti]  Last offered spring 2019.   
494       100                       [Sarah Cannon]                 Every year   
258       300                                 [Gu]                       Fall   
2848        0                                   []                              
2596      100                     [Roberto Pedace]                              
561       100                       [George Batta]                 Every year   
107       300  [Melissa E. O'Neill, Erin Talvitie]            Fall and Spring   
504       100                  [Charles Griffiths]                 Every year   
941       100                                   []                 Every year   

                                          prerequisites corequisites  \
507                                                                    
1303  BIOL 040 PO and one of the following CSCI 005 ...                
494                                                                    
258   CSCI070 HM  and (CSCI140 HM  or MATH131 HM  or...                
2848                                                                   
2596                           ECON 101  and ECON 120 .                
561                                                                    
107   (CSCI060 HM  or CSCI042 HM ), and at least one...                
504                                                                    
941                                                                    

      currently_offered  fee     score  
507                True    0  0.491876  
1303               True    0  0.448396  
494                True    0  0.436876  
258               False    0  0.393496  
2848              False    0  0.389442  
2596               True    0  0.372168  
561                True    0  0.367690  
107                True    0  0.364894  
504                True    0  0.345945  
941               False    0  0.344443

In [10]:
tfidf_word('culture', data).sort_values(by=["score"], ascending = False).head(10)

title    identifier  \
2756  Introduction to the Philosophy and History of ...   HIST-123-SC   
2785  Introduction to the Philosophy and History of ...   HMSC-123-SC   
878                               Film and Mass Culture    LIT-138-CM   
762      Culture and Society in Weimar and Nazi Germany  HIST-139E-CM   
3572                                    Popular Culture     MS-125-PZ   
2631  Literature and Popular Culture in the Antebell...  ENGL-143S-SC   
3236                      Visual Culture at the Margins   ASAM-171-PZ   
726                                    Cold War America   HIST-099-CM   
741                                    Cold War America   HIST-118-CM   
2789                 Critical Theory and Modern Culture   HMSC-136-SC   

                                            description            source  \
2756  This course will focus on some of the major wo...           Scripps   
2785  This course will focus on some of the major wo...           Scripps   
878   This course will examine film as art and as me...  ClaremontMckenna   
762   A study of the transformation of German cultur...  ClaremontMckenna   
3572  This course will cover a broad range of histor...            Pitzer   
2631  The years preceding the Civil War saw both the...           Scripps   
3236  This course will examine various forms of visu...            Pitzer   
726   The Cold War dramatically shaped the politics,...  ClaremontMckenna   
741   The Cold War dramatically shaped the politics,...  ClaremontMckenna   
2789  This course explores historical and contempora...           Scripps   

      credits     instructors           offered prerequisites corequisites  \
2756      100              []        Every year                              
2785      100              []                                                
878       100              []  Every third year                              
762       100              []  Every other year                              
3572        0              []                                                
2631      100              []  Every other year                              
3236        0              []                                                
726       100  [Lily Geismer]      Occasionally                              
741       100              []  Every other year                              
2789      100              []                                                

      currently_offered  fee     score  
2756              False    0  0.365261  
2785              False    0  0.361819  
878               False    0  0.297211  
762               False    0  0.258315  
3572              False    0  0.246987  
2631              False    0  0.245117  
3236              False    0  0.238023  
726                True    0  0.233698  
741               False    0  0.233146  
2789              False    0  0.230596

In [11]:
tfidf_word('activism', data).sort_values(by=["score"], ascending = False).head(10)

title   identifier  \
1339  Latinx Social Movements: Identity, Power, and ...  CHST-136-CH   
3156  Political Activism, Social Movements and the P...  ANTH-138-PZ   
2482                          Art, Activism, Propaganda  ARHI-184-SC   
352                         Activism, Vocation, Justice  RLST-168-HM   
3226             Introduction to Asian American Studies  ASAM-101-PZ   
2283                     Social and Political Movements   SOC-075-PO   
2973                                  Remaking the Self  POLI-173-SC   
2287  Los Angeles Communities: Transformations, Ineq...   SOC-114-CH   
342                       Global Environmental Politics  POST-140-HM   
2772        African American Women in the United States  HIST-171-AF   

                                            description      source  credits  \
1339  Latin/o/a/xs have historically used grassroots...      Pomona      100   
3156  By examining contemporary issues, themes, and ...      Pitzer        0   
2482  Explores the intersection of art, political ac...     Scripps      100   
352   The histories of social change activism are fi...  HarveyMudd      150   
3226  Introduction to the field of Asian American St...      Pitzer      100   
2283  Can activism from below change society and pol...      Pomona      100   
2973  How do social movements change the world by ch...     Scripps      100   
2287  Use of case study approach to explore the inte...      Pomona      100   
342   Analyzes the political dynamics driving global...  HarveyMudd      300   
2772  This course explores the distinctive and diver...     Scripps      100   

                                            instructors  \
1339                                     [A. Zimmerman]   
3156                                                 []   
2482                                                 []   
352                                             [Dyson]   
3226                                   [Rosanna Simons]   
2283                                          [C. Beck]   
2973                                                 []   
2287  [Jeffrey D. Groves, Frank Lubbock Miller, Fran...   
342                                         [Steinberg]   
2772                                                 []   

                        offered  \
1339                 Fall 2020.   
3156                              
2482               Occasionally   
352                               
3226                              
2283    Last offered fall 2018.   
2973               Occasionally   
2287  Last offered spring 2018.   
342                               
2772                              

                                          prerequisites corequisites  \
1339                                                                   
3156                                                                   
2482                                                                   
352                               Instructor permission                
3226                                                                   
2283                                                                   
2973                                                                   
2287  Any course in Chicanx-Latinx Studies or Sociology                
342                                                                    
2772                                                                   

      currently_offered  fee     score  
1339              False    0  0.481138  
3156              False    0  0.391203  
2482              False    0  0.306722  
352               False    0  0.259396  
3226               True    0  0.224519  
2283              False    0  0.220778  
2973              False    0  0.199427  
2287               True    0  0.176731  
342               False    0  0.174987  
2772              False    0  0.173748

In [12]:
tfidf_word('fiction', data).sort_values(by=["score"], ascending = False).head(10)

title    identifier  \
890                 Advanced Creative Writing    LIT-181-CM   
3837              Journalism in Latin America   SPAN-162-PZ   
813          Introduction to Creative Writing    LIT-031-CM   
2657          Introduction to Fiction Writing   ENGL-193-SC   
2658        Advanced Fiction Writing Workshop  ENGL-194S-SC   
204                          Fiction Workshop    LIT-035-HM   
587                 The Francophone Caribbean   FREN-115-CM   
2254  Post-Soviet Russian Culture and Society   RUSS-182-PO   
892                  Advanced Fiction Writing    LIT-183-CM   
864                European Modernist Fiction    LIT-122-CM   

                                            description            source  \
890   This is a class for the student who is serious...  ClaremontMckenna   
3837   Better than Fiction: Journalism in Latin America            Pitzer   
813   This course offers the chance to explore three...  ClaremontMckenna   
2657  This is an introductory course on writing shor...           Scripps   
2658  This advanced fiction workshop is intended for...           Scripps   
204   This course is designed as an introductory wor...        HarveyMudd   
587   A study of works of writers and artists from H...  ClaremontMckenna   
2254  The course explores the major changes in Russi...            Pomona   
892   This advanced fiction workshop is intended for...  ClaremontMckenna   
864   The first half of the 20th century produced an...  ClaremontMckenna   

      credits            instructors          offered prerequisites  \
890       100                     []       Every year                 
3837        0                     []                                  
813       100                     []     Occasionally                 
2657      100       [Leila Mansouri]                                  
2658        0                     []                                  
204       300  [Salvador Plascencia]  Fall and Spring                 
587       100                     []     Occasionally                 
2254      100    [Larissa V. Rudova]     Spring 2022.   RUSS 044 PO   
892       100       [Mary Gaitskill]       Every year                 
864       100                     []     Occasionally                 

     corequisites  currently_offered  fee     score  
890                            False    0  0.399308  
3837                           False    0  0.371970  
813                            False    0  0.364573  
2657                            True    0  0.286207  
2658                           False    0  0.277028  
204                             True    0  0.276258  
587                            False    0  0.274141  
2254                            True    0  0.257522  
892                             True    0  0.254716  
864                            False    0  0.239155

In [13]:
tfidf_word('environment', data).sort_values(by=["score"], ascending = False).head(10)

title    identifier  \
1693  Food and the Environment in Asia and the Pacific  HIST-101F-PO   
1016                              Environmental Ethics   PHIL-187-CM   
442                                       Microbiology  BIOL-168L-KS   
3379                   Critical Environmental Analysis     EA-150-PZ   
2304   Africa, the Environment, and the Global Economy   SOC-189H-PO   
3362                                     Urban Ecology     EA-098-PZ   
3135                     Global Environmental Conflict   ANTH-082-PZ   
3121           Native Americans and Their Environments   ANTH-012-PZ   
314                                  Topics in Physics   PHYS-080-HM   
1014                              Topics in Aesthetics   PHIL-184-CM   

                                            description            source  \
1693  A single question inspired this seminar: what ...            Pomona   
1016  An exploration of human beings’ ethical relati...  ClaremontMckenna   
442   In this fundamental microbiology course we wil...  ClaremontMckenna   
3379  A seminar examination of how environmental iss...            Pitzer   
2304  Drawing on sociology and related disciplines, ...            Pomona   
3362  Urban ecology is a subfield of ecology that de...            Pitzer   
3135  This class uses the tools of anthropology and ...            Pitzer   
3121  This course will investigate the traditional i...            Pitzer   
314   An area of physics is studied, together with i...        HarveyMudd   
1014  Aesthetics is the philosophical study of art a...  ClaremontMckenna   

      credits         instructors                    offered prerequisites  \
1693      100      [S. Yamashita]  Last offered spring 2018.                 
1016      100                  []               Occasionally                 
442       100  [Pete Chandrangsu]               Occasionally                 
3379        0                  []                                            
2304      100       [S. Stefanos]                 Each fall.                 
3362      100  [Heather Campbell]                                            
3135        0                  []                                            
3121      100     [Sheryl Miller]                                            
314       300   [Donnelly, Saeta]                              PHYS051 HM    
1014      100                  []               Occasionally                 

     corequisites  currently_offered  fee     score  
1693                           False    0  0.339170  
1016                           False    0  0.286202  
442                             True    0  0.269543  
3379                           False    0  0.256764  
2304                           False    0  0.221139  
3362                            True    0  0.206209  
3135                           False    0  0.204906  
3121                            True    0  0.199003  
314                            False    0  0.195351  
1014                           False    0  0.192540

### Rank classes based on a given class
`tfidf_id(id, data)` will take in a identifier of some course (i.e. "PHIL-187-CM") and return courses that are most similar to the givern course. To do it, we will look at which words ranked highest for the given class and find other classes where same word also ranked highest.

Similarly to `tfidf_word`, we will add new columns that have the score of each class based on each word. In thins case we will have multiple new columns that have similarity scores for different word that were important in our input class.

Additionally, if `tfidf_id` is used with `penalize=True` the courses withing the same department will be penalized

In [14]:
def tfidf_word_helper(word, data1, prefix=""):
    data2 = data1.copy()
    corpus = list(data2.description)
    vectorizer = TfidfVectorizer(use_idf=True)
    vectors = vectorizer.fit_transform(corpus)

    score_for_word = []
    words = vectorizer.get_feature_names()
    try:
        index = words.index(word)
    except:
        print("'" + word + "'" + " is not mentioned in any course descriptions")
        return

    for i in range(0, len(corpus)):
        value = vectors[i].T.todense()[index]
        score_for_word.append(value)

    score_for_word = [float(i) for i in score_for_word] # type cast each score to a float

    data2["score"+prefix] = score_for_word
    return data2


def tfidf_id(id, data_):
    index_id = list(data_.index[data_["identifier"] == id]) # index of input class
    
    if len(index_id) == 0:
        print("Couldn't find a course " + id)
        return
    else:
        index_id = index_id[0]
    print(data_.loc[index_id, "identifier"])
    print(data_.loc[index_id, "description"])

    corpus = list(data_.description)
    vectorizer = TfidfVectorizer(use_idf=True)
    vectors = vectorizer.fit_transform(corpus)
    words = vectorizer.get_feature_names()

    scores_id = vectors[index_id].T.todense() # score values for our given class
    scores_id = [float(i) for i in scores_id] # type cast to float
    score_for_word = {}

    
    for i in range(0, len(words)):
        if scores_id[i] > 0.2: # We will count a word as relevant if it's score if more than 0.2 (arbitrary value, subject to change)
            score_for_word[words[i]] = [scores_id[i], i]
    score_for_word = {k: v for k, v in sorted(score_for_word.items(), key=lambda item: item[1][0], reverse=True)} #ordear dict by descending vals
    print("Most important words in course "+ id + " and their scores and indices")
    print(score_for_word)

    data__ = data_.copy() ## since we're making changes to our dataframe, we don't want to save these changes in the original dataframe
    new_cols = []
    for word in score_for_word:
        new_cols.append("score_" + word)
        data__ = tfidf_word_helper(word, data__, "_"+word) # tf idf word helper will help us add new columns with prefixes, same algo as regular tfdf_word()

    data__ = data__.sort_values(by=new_cols, ascending= False)
    return data__

In [15]:
def tfidf_id(id, data_, penalty=1.0):
    index_id = list(data_.index[data_["identifier"] == id]) # index of input class
    
    if len(index_id) == 0:
        print("Couldn't find a course " + id)
        return
    else:
        index_id = index_id[0]
    #print(data_.loc[index_id, "identifier"])
    #print(data_.loc[index_id, "description"])

    corpus = list(data_.description)
    vectorizer = TfidfVectorizer(use_idf=True)
    vectors = vectorizer.fit_transform(corpus)
    words = vectorizer.get_feature_names()

    scores_id = vectors[index_id].T.todense() # score values for our given class
    scores_id = [float(i) for i in scores_id] # type cast to float
    score_for_word = {}

    
    for i in range(0, len(words)):
        if scores_id[i] > 0.2: # We will count a word as relevant if it's score if more than 0.2 (arbitrary value, subject to change)
            score_for_word[words[i]] = [scores_id[i], i]
    score_for_word = {k: v for k, v in sorted(score_for_word.items(), key=lambda item: item[1][0], reverse=True)} #ordear dict by descending vals
    #print("Most important words in course "+ id + " and their scores and indices")
    #print(score_for_word)

    word = list(score_for_word.keys())[0]
    print("Giving course recomendations based of the word: " + word)

    data__ = data_.copy() ## since we're making changes to our dataframe, we don't want to save these changes in the original dataframe
    data__ = tfidf_word(word, data__)


    if penalty!=1.0:
        department = id[:id.index("-")]
        print("penalizing " + department +" courses by a factor of " + str(penalty))
        
        for index in data__.index:
            if data__["identifier"][index] == id:
                data__["score"][index] = 0.0
            elif department in data__["identifier"][index]:
                data__["score"][index] = data__["score"][index]*penalty
    else:
        print("No penalty")

    return data__

In [16]:
tfidf_id("AMST-120-HM", data).sort_values(by=["score"], ascending = False).head(5)

Giving course recomendations based of the word: immigrants
No penalty


title   identifier  \
2129        The Politics of Immigration and Citizenship  POLI-046-PO   
2                                  Hyphenated Americans  AMST-120-HM   
2808  Italians as Guests and Hosts: Intercultural En...  ITAL-136-SC   
1151  Images of Immigration in Spanish Literature an...  SPAN-122-CM   
3402                      Criminalization of Immigrants    FS-016-PZ   

                                            description            source  \
2129  Examines immigration and citizenship politics ...            Pomona   
2     A focus on the experience of immigrants in the...        HarveyMudd   
2808  This course examines the phenomenon of exchang...           Scripps   
1151  From an interdisciplinary perspective, this co...  ClaremontMckenna   
3402  How did immigration and the U.S. - Mexico bord...            Pitzer   

      credits            instructors                    offered  \
2129      100                [Staff]  Last offered spring 2019.   
2         300             [Balseiro]                              
2808      100                     []                              
1151      100                     []           Every other year   
3402      100  [Steffanie Guillermo]                              

                 prerequisites corequisites  currently_offered  fee     score  
2129                                                     False    0  0.325699  
2                                                        False    0  0.270499  
2808  ITAL 044  or equivalent.                           False    0  0.238728  
1151                                                     False    0  0.211420  
3402                                                      True    0  0.175879

In [17]:
tfidf_id("AMST-120-HM", data, penalty=0.5).sort_values(by=["score"], ascending = False).head(5)

Giving course recomendations based of the word: immigrants
penalizing AMST courses by a factor of 0.5


title   identifier  \
2129        The Politics of Immigration and Citizenship  POLI-046-PO   
2808  Italians as Guests and Hosts: Intercultural En...  ITAL-136-SC   
1151  Images of Immigration in Spanish Literature an...  SPAN-122-CM   
3402                      Criminalization of Immigrants    FS-016-PZ   
3695          US Immigration and Transnational Politics  POST-174-CH   

                                            description            source  \
2129  Examines immigration and citizenship politics ...            Pomona   
2808  This course examines the phenomenon of exchang...           Scripps   
1151  From an interdisciplinary perspective, this co...  ClaremontMckenna   
3402  How did immigration and the U.S. - Mexico bord...            Pitzer   
3695  Examines the factors shaping the size and comp...            Pitzer   

      credits            instructors                    offered  \
2129      100                [Staff]  Last offered spring 2019.   
2808      100                     []                              
1151      100                     []           Every other year   
3402      100  [Steffanie Guillermo]                              
3695        0                     []                              

                 prerequisites corequisites  currently_offered  fee     score  
2129                                                     False    0  0.325699  
2808  ITAL 044  or equivalent.                           False    0  0.238728  
1151                                                     False    0  0.211420  
3402                                                      True    0  0.175879  
3695                                                     False    0  0.169333

In [18]:
tfidf_id("CSCI-036-CM", data).sort_values(by=["score"], ascending = False).head(5)

Giving course recomendations based of the word: data
No penalty


title   identifier  \
507                   Advanced Projects in Data Science    DS-180-CM   
1303  Data Analysis and Programming for the Life Sci...  BIOL-174-PO   
494                         Foundations of Data Science  CSCI-036-CM   
258                            Nonlinear Data Analytics  MATH-178-HM   
2848                       CS1: Intro to Python and Viz    MS-059-SC   

                                            description            source  \
507   This course allows teams of students to wrestl...  ClaremontMckenna   
1303  This course explores the analysis of big data ...            Pomona   
494   Data science is the interdisciplinary study of...  ClaremontMckenna   
258   Analysis of nonlinear large dynamic data inclu...        HarveyMudd   
2848  This is an introduction to computer programmin...           Scripps   

      credits         instructors                    offered  \
507       100         [Jeho Park]                 Every year   
1303      100  [Andre Cavalcanti]  Last offered spring 2019.   
494       100      [Sarah Cannon]                 Every year   
258       300                [Gu]                       Fall   
2848        0                  []                              

                                          prerequisites corequisites  \
507                                                                    
1303  BIOL 040 PO and one of the following CSCI 005 ...                
494                                                                    
258   CSCI070 HM  and (CSCI140 HM  or MATH131 HM  or...                
2848                                                                   

      currently_offered  fee     score  
507                True    0  0.491876  
1303               True    0  0.448396  
494                True    0  0.436876  
258               False    0  0.393496  
2848              False    0  0.389442

In [19]:
tfidf_id("CSCI-036-CM", data, penalty=0.5).sort_values(by=["score"], ascending = False).head(5)

Giving course recomendations based of the word: data
penalizing CSCI courses by a factor of 0.5


title   identifier  \
507                   Advanced Projects in Data Science    DS-180-CM   
1303  Data Analysis and Programming for the Life Sci...  BIOL-174-PO   
258                            Nonlinear Data Analytics  MATH-178-HM   
2848                       CS1: Intro to Python and Viz    MS-059-SC   
2596                                       Econometrics  ECON-125-SC   

                                            description            source  \
507   This course allows teams of students to wrestl...  ClaremontMckenna   
1303  This course explores the analysis of big data ...            Pomona   
258   Analysis of nonlinear large dynamic data inclu...        HarveyMudd   
2848  This is an introduction to computer programmin...           Scripps   
2596  Statistical techniques for testing economic mo...           Scripps   

      credits         instructors                    offered  \
507       100         [Jeho Park]                 Every year   
1303      100  [Andre Cavalcanti]  Last offered spring 2019.   
258       300                [Gu]                       Fall   
2848        0                  []                              
2596      100    [Roberto Pedace]                              

                                          prerequisites corequisites  \
507                                                                    
1303  BIOL 040 PO and one of the following CSCI 005 ...                
258   CSCI070 HM  and (CSCI140 HM  or MATH131 HM  or...                
2848                                                                   
2596                           ECON 101  and ECON 120 .                

      currently_offered  fee     score  
507                True    0  0.491876  
1303               True    0  0.448396  
258               False    0  0.393496  
2848              False    0  0.389442  
2596               True    0  0.372168

In [20]:
tfidf_id("LIT-138-CM", data).sort_values(by=["score"], ascending = False).head(5)

Giving course recomendations based of the word: film
No penalty


title  identifier  \
874   Special Studies in Film  LIT-134-CM   
870      Introduction to Film  LIT-130-CM   
878     Film and Mass Culture  LIT-138-CM   
3564               Film Sound   MS-114-PZ   
285                Film Music  MUS-067-HM   

                                            description            source  \
874   A seminar designed to explore the aesthetic ac...  ClaremontMckenna   
870   We will begin with a close analysis of a conte...  ClaremontMckenna   
878   This course will examine film as art and as me...  ClaremontMckenna   
3564  An intermediate level media history and theory...            Pitzer   
285   An exploration of the history and aesthetics o...        HarveyMudd   

      credits            instructors           offered prerequisites  \
874       100  [Robert von Hallberg]  Every other year                 
870       100         [Thomas Schur]        Every year                 
878       100                     []  Every third year                 
3564        0                     []                                   
285       300                [Alves]                                   

     corequisites  currently_offered  fee     score  
874                             True    0  0.551764  
870                             True    0  0.502270  
878                            False    0  0.445085  
3564                           False    0  0.354066  
285                            False    0  0.350366

In [21]:
tfidf_id("LIT-138-CM", data, penalty=0.5).sort_values(by=["score"], ascending = False).head(5)

Giving course recomendations based of the word: film
penalizing LIT courses by a factor of 0.5


title   identifier  \
3564                                       Film Sound    MS-114-PZ   
285                                        Film Music   MUS-067-HM   
1875                                      Film Theory   MS-148G-PO   
3740                           Cognitive Film Studies  PSYC-128-PZ   
1718  Hollywood, War, and Empire: The Historical Film  HIST-129-PO   

                                            description      source  credits  \
3564  An intermediate level media history and theory...      Pitzer        0   
285   An exploration of the history and aesthetics o...  HarveyMudd      300   
1875  This course develops theoretical approaches to...      Pomona      100   
3740  This course examines the moving image from the...      Pitzer        0   
1718  Film evolved at the same time as modern global...      Pomona      100   

         instructors                    offered  \
3564              []                              
285          [Alves]                              
1875   [T. Connelly]               Spring 2019.   
3740              []                              
1718  [V. Silverman]  Last offered spring 2018.   

                             prerequisites corequisites  currently_offered  \
3564                                                                 False   
285                                                                  False   
1875  MS 049 PO , MS 050 PO , or MS 051 PO                           False   
3740                                                                 False   
1718                                                                 False   

      fee     score  
3564    0  0.354066  
285     0  0.350366  
1875    0  0.341641  
3740    0  0.339891  
1718    0  0.314672